### PR TITLE
feat(attribute): add map to attribute

### PR DESCRIPTION
**Pull request by joelwurtz**
Add `#[MapTo]` attribute which allows to : 

 * Add a mapping on a previously not found property
 * Ignore a property during a Map ref #16 
 * Customize the target property ref #34 
 * Allow to use a custom transformer : callable / transformer service ref #28

### CODE REVIEW

File: src/EventListener/MapToListener.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+use AutoMapper\Transformer\CustomTransformer\CustomModelTransformerInterface;
+use AutoMapper\Transformer\CustomTransformer\CustomPropertyTransformer;
+use AutoMapper\Transformer\CustomTransformer\CustomPropertyTransformerInterface;
+use AutoMapper\Transformer\CustomTransformer\CustomTransformersRegistry;
+
+final readonly class MapToListener
+{
+    public function __construct(private CustomTransformersRegistry $customTransformersRegistry)
+    {
+    }
```
Comment (Korbeil) [inline src/EventListener/MapToListener.php:24]:
```suggestion
    public function __construct(
        private CustomTransformersRegistry $customTransformersRegistry
    ) {
    }
```

File: src/EventListener/MapToListener.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        if ($event->mapperMetadata->sourceReflectionClass === null) {
+            return;
+        }
+
+        $properties = $event->mapperMetadata->sourceReflectionClass->getProperties();
+
+        foreach ($properties as $reflectionProperty) {
+            $mapToAttributes = $reflectionProperty->getAttributes(MapTo::class);
+
+            if (empty($mapToAttributes)) {
```
Comment (Korbeil) [inline src/EventListener/MapToListener.php:37]:
```suggestion
            if (0 === \count($mapToAttributes)) {
```

File: src/EventListener/MapToListener.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                continue;
+            }
+
+            foreach ($mapToAttributes as $mapToAttribute) {
```
Comment (Korbeil) [inline src/EventListener/MapToListener.php:41]: Do we have some kind of priority between attributes there ? Or it's first written onto the class that is first here ?

File: src/EventListener/MapToListener.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+                    if (\is_string($mapToAttributeInstance->transformer) && $customTransformer = $this->customTransformersRegistry->getCustomTransformer($mapToAttributeInstance->transformer)) {
+                        if ($customTransformer instanceof CustomModelTransformerInterface) {
+                            $transformer = new CustomModelTransformer($mapToAttributeInstance->transformer);
+                        }
+
+                        if ($customTransformer instanceof CustomPropertyTransformerInterface) {
+                            $transformer = new CustomPropertyTransformer($mapToAttributeInstance->transformer);
+                        }
+                    } elseif (@\is_callable($mapToAttributeInstance->transformer, false, $callableName) && $callableName !== null) {
```
Comment (Korbeil) [inline src/EventListener/MapToListener.php:65]: Is that `@` required ?

File: src/EventListener/MapToListener.php [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                $property = new PropertyMetadataEvent(
+                    $event->mapperMetadata,
+                    $sourceProperty,
+                    $targetProperty,
+                    $mapToAttributeInstance->maxDepth,
+                    $transformer,
+                    $mapToAttributeInstance->ignore,
+                );
+
+                if (isset($event->properties[$property->target->name])) {
```
Comment (Korbeil) [inline src/EventListener/MapToListener.php:79]:
```suggestion
                if (array_key_eixsts($property->target->name, $event->properties)) {
```

File: src/Transformer/CallableTransformer.php [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+use AutoMapper\Generator\UniqueVariableScope;
+use AutoMapper\Metadata\PropertyMetadata;
+use PhpParser\Node\Arg;
+use PhpParser\Node\Expr;
+use PhpParser\Node\Scalar;
+
+class CallableTransformer implements TransformerInterface
+{
+    public function __construct(private string $callable)
+    {
```
Comment (Korbeil) [inline src/Transformer/CallableTransformer.php:16]:
```suggestion
    public function __construct(
        private string $callable
    ) {
```

File: src/Transformer/CallableTransformer.php [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+use AutoMapper\Generator\UniqueVariableScope;
+use AutoMapper\Metadata\PropertyMetadata;
+use PhpParser\Node\Arg;
+use PhpParser\Node\Expr;
+use PhpParser\Node\Scalar;
+
+class CallableTransformer implements TransformerInterface
+{
+    public function __construct(private string $callable)
```
Comment (Korbeil) [inline src/Transformer/CallableTransformer.php:15]: Is that because we can't handle "real" callable in attributes ? (things like `fn($var) => $var` for exemple ?

File: src/EventListener/MapToListener.php (hunk 3 above)
Comment (Korbeil) [reply src/EventListener/MapToListener.php:41]: No reason to have priority since we throw an exception when we have a rule for the same target.

File: src/Transformer/CallableTransformer.php (hunk 7 above)
Comment (joelwurtz) [reply src/Transformer/CallableTransformer.php:15]: Yes otherwise it's "impossible" to generate (we can use the previous callback extractor but imo this is a rabbithole)

**Review by Korbeil — APPROVED** [review]
LGTM

File: src/EventListener/MapToListener.php (hunk 4 above)
Comment (joelwurtz) [reply src/EventListener/MapToListener.php:66]: Not really but current support is not great, we should do like the callback validator of symfony in term of support without the `Closure`

File: src/EventListener/MapToListener.php [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+
+    private function setPropertiesFromSource(GenerateMapperEvent $event): void
+    {
+        if ($event->mapperMetadata->sourceReflectionClass === null) {
+            return;
+        }
+
+        $properties = $event->mapperMetadata->sourceReflectionClass->getProperties();
+
+        foreach ($properties as $reflectionProperty) {
```
Comment (nikophil) [inline src/EventListener/MapToListener.php:51]: I think this method (and `setPropertiesFromTarget()`) can be simplified by looping over
`[...$properties, ...$methods]`

File: src/EventListener/MapToListener.php [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+
+        $this->setPropertiesFromSource($event);
+        $this->setPropertiesFromTarget($event);
+    }
+
+    private function setPropertiesFromSource(GenerateMapperEvent $event): void
+    {
+        if ($event->mapperMetadata->sourceReflectionClass === null) {
+            return;
+        }
```
Comment (nikophil) [inline src/EventListener/MapToListener.php:47]: seems irrelevant, because this condition is already in `__invoke()` method

File: src/EventListener/MapToListener.php [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+    }
+
+    public function __invoke(GenerateMapperEvent $event): void
+    {
+        if ($event->mapperMetadata->sourceReflectionClass === null) {
+            return;
+        }
+
+        $this->setPropertiesFromSource($event);
+        $this->setPropertiesFromTarget($event);
```
Comment (nikophil) [inline src/EventListener/MapToListener.php:40]: shouldn't this `MapToListener` only handle `MapTo` attributes? and handle `MapFrom` in another listener?
Here, because of the previous condition, we cannot deal with "array to object" transformations (I don't know if you are planning to do this in another PR)

File: src/EventListener/MapToListener.php [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+        private CustomTransformersRegistry $customTransformersRegistry,
+        private InflectorInterface $inflector = new EnglishInflector(),
+    ) {
+    }
+
+    public function __invoke(GenerateMapperEvent $event): void
+    {
+        if ($event->mapperMetadata->sourceReflectionClass === null) {
+            return;
+        }
```
Comment (nikophil) [inline src/EventListener/MapToListener.php:37]: here is an idea we could use all over the code to improve readability: I think we should introduce an enum which would looks like:
```php
enum MappingType
{
    case FROM_SOURCE; // (aka NORMALIZATION)
    case FROM_TARGET; // (aka DENORMALIZATION)
    case SOURCE_TARGET; // (aka OBJECTS_MAPPING)
}
```
and rely on this enum instead of relying on this kind of condition which is kinda meaningless

File: src/Transformer/CallableTransformer.php [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace AutoMapper\Transformer;
+
+use AutoMapper\Generator\UniqueVariableScope;
+use AutoMapper\Metadata\PropertyMetadata;
+use PhpParser\Node\Arg;
+use PhpParser\Node\Expr;
+use PhpParser\Node\Scalar;
+
+class CallableTransformer implements TransformerInterface
```
Comment (nikophil) [inline src/Transformer/CallableTransformer.php:13]: final / readonly / internal ?

File: tests/Fixtures/MapTo/Bar.php [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+declare(strict_types=1);
+
+namespace AutoMapper\Tests\Fixtures\MapTo;
+
+use AutoMapper\Attribute\MapFrom;
+
+class Bar
+{
+    #[MapFrom(source: 'array', ignore: true)]
```
Comment (nikophil) [inline tests/Fixtures/MapTo/Bar.php:11]: does this work?
